ambry-store/src/test/java/com.github.ambry.store/BlobStoreStatsTest.java

```diff
@@ -0,0 +1,335 @@
+   */
+  @Test
+  public void testValidDataSizeAfterExpiration() throws InterruptedException, StoreException, IOException {
+    // advance time by 1 second for previous deletes/expiration to take effect
+    state.advanceTime(Time.MsPerSec);
+    long timeInMsBeforeExpiringPuts = state.time.milliseconds();
+    long totalLogSegmentValidSizeBeforePuts =
+        verifyLogSegmentValidSize(new TimeRange(timeInMsBeforeExpiringPuts, 0L));
+    long totalContainerValidSizeBeforePuts = verifyContainerValidSize();
+
+    // 3 puts that will expire in 20 seconds
+    state.addPutEntries(3, CuratedLogIndexState.PUT_RECORD_SIZE, timeInMsBeforeExpiringPuts + 20 * Time.MsPerSec);
+    // advance time by 21 seconds for expiration to take effect
+    state.advanceTime(21 * Time.MsPerSec);
+
+    long timeInMsAfterExpiringPuts = state.time.milliseconds();
+    long totalLogSegmentValidSizeAfterPuts = verifyLogSegmentValidSize(new TimeRange(timeInMsAfterExpiringPuts, 0L));
+    long totalContainerValidSizeAfterPuts = verifyContainerValidSize();
+    assertEquals("Expired put entries are not properly counted for log segment valid size",
+        totalLogSegmentValidSizeBeforePuts, totalLogSegmentValidSizeAfterPuts);
+    assertEquals("Expired put entries are not properly counted for container valid size",
+        totalContainerValidSizeBeforePuts, totalContainerValidSizeAfterPuts);
+  }
+
+  /**
+   * Tests to verify the correctness of reported stats with after new deletes via the following steps:
+   * 1. Add new puts that are going to be deleted later.
+   * 2. Verify reported stats and record the total valid size before new deletes.
+   * 3. Perform the deletes.
+   * 4. Verify reported stats and record the total valid size after the deletes.
+   * 5. Verify the delta of total valid size prior to the new deletes and after matches with the expected delta.
+   * @throws InterruptedException
+   * @throws StoreException
+   * @throws IOException
+   */
+  @Test
+  public void testValidDataSizeAfterDeletes() throws InterruptedException, StoreException, IOException {
+    // 5 puts to be deleted
+    state.addPutEntries(5, CuratedLogIndexState.PUT_RECORD_SIZE, Utils.Infinite_Time);
+
+    // advance time by 1 second for deletes/expiration to take effect
+    state.advanceTime(Time.MsPerSec);
+    long timeInMsBeforeDeletes = state.time.milliseconds();
+    long totalLogSegmentValidSizeBeforeDeletes = verifyLogSegmentValidSize(new TimeRange(timeInMsBeforeDeletes, 0L));
+    long totalContainerValidSizeBeforeDeletes = verifyContainerValidSize();
+
+    // 2 deletes from the last index segment
+    state.addDeleteEntry(state.getIdToDeleteFromIndexSegment(state.referenceIndex.lastKey()));
+    state.addDeleteEntry(state.getIdToDeleteFromIndexSegment(state.referenceIndex.lastKey()));
+    // advance time by 1 second for deletes/expiration to take effect
+    state.advanceTime(Time.MsPerSec);
+
+    long timeInMsAfterDeletes = state.time.milliseconds();
```

Comment:
Checking at a time point before this time can only verify against CuratedLogIndexState. There is no delta we can check unless we have the knowledge if the new deletes triggered a roll over or not.